jpa/eclipselink.jpa.test/src/org/eclipse/persistence/testing/tests/jpa/advanced/PersistenceUnitProcessorTest.java

```diff
@@ -35,42 +35,140 @@
+        checkPURootCustom(
+            "zip", "/foo/bar.jar!/foo/META-INF/persistence.xml",
+            "file:/foo/bar.jar"
+        );
+
+        // Test cases for specific issues.
+
+        // 463571
+
+        checkPURootCustom(
+            "zip", "/C:/Oracle/Middleware/Oracle_Home/wlserver/samples/domains/mydomain/servers/myserver/tmp/_WL_user/eclipselink-advanced-model/y986my/eclipselink-advanced-model_ejb.jar!/META-INF/persistence.xml",
+            "file:/C:/Oracle/Middleware/Oracle_Home/wlserver/samples/domains/mydomain/servers/myserver/tmp/_WL_user/eclipselink-advanced-model/y986my/eclipselink-advanced-model_ejb.jar"
+        );
+
+        checkPURootCustom(
+            "zip", "/C:/Program Files/Middleware/Oracle_Home/wlserver/samples/domains/mydomain/servers/myserver/tmp/_WL_user/eclipselink-advanced-model/y986my/eclipselink-advanced-model_ejb.jar!/META-INF/persistence.xml",
+            "file:/C:/Program Files/Middleware/Oracle_Home/wlserver/samples/domains/mydomain/servers/myserver/tmp/_WL_user/eclipselink-advanced-model/y986my/eclipselink-advanced-model_ejb.jar"
+        );
+
+        checkPURootCustom(
+            "zip", "/C:/Program.Files/Middleware/Oracle_Home/wlserver/samples/domains/mydomain/servers/myserver/tmp/_WL_user/eclipselink-advanced-model/y986my/eclipselink-advanced-model_ejb.jar!/META-INF/persistence.xml",
+            "file:/C:/Program.Files/Middleware/Oracle_Home/wlserver/samples/domains/mydomain/servers/myserver/tmp/_WL_user/eclipselink-advanced-model/y986my/eclipselink-advanced-model_ejb.jar"
+        );
+    }
+
+    public void testComputePURootURLForJarFile() throws Exception {
+        // Test cases for expected behavior.
+
+        checkPURootSimple(
+            "jar:file:/foo/bar.jar!/META-INF/persistence.xml",
+            "file:/foo/bar.jar"
+        );
+
+        checkPURootSimple(
+            "jar:file:/foo/bar.war!/WEB-INF/classes/META-INF/persistence.xml",
+            "jar:file:/foo/bar.war!/WEB-INF/classes/"
+        );
+
+        checkPURootSimple(
+            "jar:file:/foo/bar.jar!/WEB-INF/classes/META-INF/persistence.xml",
+            "file:/foo/bar.jar"
+        );
+
+        checkPURootSimple(
+            "jar:file:/foo/bar.jar!/foo/META-INF/persistence.xml",
+            "file:/foo/bar.jar"
+        );
+    }
+
+    public void testComputePURootURLForWsjarFile() throws Exception {
+        // Test cases for expected behavior.
+        // The results differ slightly from the other archive URLs!
+
+        checkPURootCustom(
+            "wsjar", "file:/foo/bar.jar!/META-INF/persistence.xml",
+            "jar:file:/foo/bar.jar!/"
+        );
+
+        checkPURootCustom(
+            "wsjar", "file:/foo/bar.war!/WEB-INF/classes/META-INF/persistence.xml",
+            "jar:file:/foo/bar.war!/WEB-INF/classes/"
+        );
+
+        checkPURootCustom(
+            "wsjar", "file:/foo/bar.jar!/WEB-INF/classes/META-INF/persistence.xml",
+            "jar:file:/foo/bar.jar!/"
+        );
```

Comment:
I'm fairly certain this is not a usecase covered by Section 8.2:

    In Java EE environments, the root of a persistence unit must be one of the following:
     • an EJB-JAR file
     • the WEB-INF/classes directory of a WAR file[87]
     • a jar file in the WEB-INF/lib directory of a WAR file
     • a jar file in the EAR library directory
     • an application client jar file

Reply:
This isnt a negative test. A negative test would be validating that passing a bad URL will result in a failure. This test is verifying that the URL ```bar.jar!/WEB-INF/classes/META-INF/persistence.xml``` gives us back the expected result of ```jar:file:/foo/bar.jar!/``` The URL is invalid in the first place. We should give back like null or throw an exception or something. Giving a result back makes it look like this is acceptable.
Granted we want to preserve

Reply:
Got it, will do that tomorrow.